pkg/controller/operators/decorators/operator.go

```diff
@@ -0,0 +1,277 @@
+package decorators
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	operatorsv2alpha1 "github.com/operator-framework/api/pkg/operators/v2alpha1"
+	"github.com/yalp/jsonpath"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/reference"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/codec"
+)
+
+const (
+	newOperatorError            = "Cannot create new Operator: %s"
+	newComponentError           = "Cannot create new Component: %s"
+	componentLabelKeyError      = "Cannot generate component label key: %s"
+	componentConditionsJSONPath = "$.status.conditions"
+
+	// ComponentLabelKeyPrefix is the key prefix used for labels marking operator component resources.
+	ComponentLabelKeyPrefix = "operators.coreos.com/"
+)
+
+// OperatorNames returns a list of operator names extracted from the given labels.
+func OperatorNames(labels map[string]string) (names []types.NamespacedName) {
+	for key := range labels {
+		if !strings.HasPrefix(key, ComponentLabelKeyPrefix) {
+			continue
+		}
+
+		names = append(names, types.NamespacedName{
+			Name: strings.TrimPrefix(key, ComponentLabelKeyPrefix),
+		})
+	}
+
+	return
+}
+
+type OperatorFactory interface {
```

Comment:
nit: naming here is getting a bit confusing for the callsites, right? `operator := factory.NewOperator(operator)`?